### ISSUE TITLE
any plans to have HMR / auto refresh in sandbox?

### ISSUE THREAD

**Issue by YarivGilad** [architect/architect#623]
any plans to have HMR / auto refresh in sandbox?

**Comment by brianleroux** [comment]
Absolutely, some of that code is written but still a WIP https://github.com/architect/sandbox/blob/c9daeb9e74294e68a13fe2d445faf6775e66c215/src/cli/index.js#L83

**Comment by ryanblock** [comment]
That line of code is more specifically about starting up new services in Sandbox (example: project doesn't have `@http` or `@static`, user adds `@http` to `.arc`, Sandbox auto-starts up http server without needing to be restarted).

I assume this is more about auto refresh in the browser?

If so, I've thought about it, but haven't yet had the time to really investigate. How should it work?

**Comment by brianleroux** [comment]
Fixed! Been available in sandbox for some time now.

**Comment by ryanblock** [comment]
Left this open, I don't thing we properly support HMR yet!

**Comment by filmaj** [comment]
> If so, I've thought about it, but haven't yet had the time to really investigate. How should it work?

I think bundlers like parcel that support HMR work by injecting a websocket listener into the client and running a small websocket server that will push a refresh event to the client when the server detects changes.

Digging around the parcel code base, I was able to find [an HMR server implementation](https://github.com/parcel-bundler/parcel/blob/v2/packages/core/parcel-bundler/src/HMRServer.js) as well as the [HMR "runtime"](https://github.com/parcel-bundler/parcel/tree/v2/packages/runtimes/hmr) which I _believe_ gets injected into the client in some fashion

**Comment by brianleroux** [comment]
FWIW you cannot reload esmodules. (The graph is static.) So… you have to reload the tree which you do by reloading the page. Unless we want to recreate esm with a bundler which…I don't think we do.
